test/integration/tracked_watermark_buffer.h

```diff
@@ -60,19 +81,86 @@ class TrackedWatermarkBufferFactory : public Buffer::WatermarkFactory {
   // functionality is disabled.
   std::pair<uint32_t, uint32_t> highWatermarkRange() const;
 
+  // Total bytes currently buffered across all known buffers.
+  uint64_t totalBytesBuffered() const {
+    absl::MutexLock lock(&mutex_);
+    return total_buffer_size_;
+  }
+
+  // Wait until total bytes buffered exceeds the a given size.
+  bool
+  waitUntilTotalBufferedExceeds(uint64_t byte_size,
+                                std::chrono::milliseconds timeout = TestUtility::DefaultTimeout);
+
+  // Set the expected account balance, prior to sending requests.
+  // The test thread can then wait for this condition to be true.
+  // This is separated so that the test thread can spin up requests however it
+  // desires in between.
+  //
+  // The Envoy worker thread will notify the test thread once the condition is
+  // met.
+  void setExpectedAccountBalance(uint64_t byte_size_per_account, uint32_t num_accounts);
+  bool waitForExpectedAccountBalanceWithTimeout(
+      std::chrono::milliseconds timeout = TestUtility::DefaultTimeout);
+
+  // Wait for the expected number of accounts and number of bound buffers.
+  //
+  // Due to deferred deletion, it possible that the Envoy hasn't cleaned up on
+  // its end, but the stream has been completed. This avoids that by awaiting
+  // for the side effect of the deletion to have occurred.
+  bool waitUntilExpectedNumberOfAccountsAndBoundBuffers(
+      uint32_t num_accounts, uint32_t num_bound_buffers,
+      std::chrono::milliseconds timeout = TestUtility::DefaultTimeout);
+
+  using AccountToBoundBuffersMap =
+      absl::flat_hash_map<BufferMemoryAccountSharedPtr,
+                          absl::flat_hash_set<TrackedWatermarkBuffer*>>;
+  void inspectAccounts(std::function<void(AccountToBoundBuffersMap&)> func) {
+    absl::MutexLock lock(&mutex_);
+    func(account_infos_);
```

Comment:
Should inspectAccounts run the inspection function in the worker thread instead, and use a notification to wait for execution to succeed before continuing?

The idea behind this is that a callback running on the worker thread would be allowed to access account methods and query buffers about their size and other useful properties.

Reply:
Done.